passes/pmgen/split_shiftx.pmg

```diff
@@ -0,0 +1,59 @@
+state <SigSpec> shiftxB
+
+match shiftx
+	select shiftx->type == $shiftx
+	select param(shiftx, \Y_WIDTH).as_int() > 1
+endmatch
+
+match macc
+	select macc->type == $macc
+	select param(macc, \B_WIDTH).as_int() == 0
+	optional
+endmatch
+
+code shiftxB
+	shiftxB = port(shiftx, \B);
+
+	if (macc) {
+		const int b_width = param(shiftx, \B_WIDTH).as_int();
+		if (param(shiftx, \B_SIGNED) != 0 && shiftxB[b_width-1] == RTLIL::S0)
+			shiftxB = shiftxB.extract(0, b_width-1);
+
+		if (port(macc, \Y) != shiftxB) {
+			blacklist(shiftx);
+			reject;
```

Comment:
I don't understand what you are trying to do here. You are blacklisting this shiftx when it happens to not match the one macc cell you just tested? What if it matches another macc cell in the module? You'll never know, because this will backtrack and just select the next shiftx.

Reply:
This is my hacky way of communicating to my pass that we should ignore this entire pattern, this pattern being that:
1. We have a `$shiftx`
2. That `$shiftx` is being driven by a `$macc`.
3. But that `$macc` doesn't fulfill all conditions (in this case, Y does not drive a potentially-sign-bit-less `B`)
As I said in the description, I'm not sure how I can express "give up on this `$shiftx` pattern", if that makes more sense now.